languages/csharp/AppEncryption/AppEncryption.Tests/AppEncryption/SessionFactoryTest.cs

```diff
@@ -56,13 +56,682 @@ private void TestConstructor()
+                ThreadPool.GetMinThreads(out _, out var currentMinIOC);
+                Assert.True(ThreadPool.SetMinThreads(numThreads, currentMinIOC));
+
+                long completedTasks = 0;
+                Parallel.ForEach(Enumerable.Range(0, numRequests), i =>
+                {
+                    using (Session<byte[], byte[]> sessionBytes = factory.GetSessionBytes(TestPartitionId))
+                    {
+                        byte[] payload = { 0, 1, 2, 3, 4, 5, 6, 7 };
+                        byte[] drr = sessionBytes.Encrypt(payload);
+                        byte[] decryptedPayload = sessionBytes.Decrypt(drr);
+
+                        Assert.Equal(payload, decryptedPayload);
+                        Interlocked.Increment(ref completedTasks);
+
+                        // Verify that cache does not have more entries than max size
+                        Assert.True(factory.SessionCache.Count <= sessionCacheMaxSize);
+                    }
+                });
+
+                // Wait for all threads to complete
+                Assert.Equal(numRequests, completedTasks);
+            }
+        }
+
+        [Fact]
+        private void TestSessionCacheMultiThreadedWithMaxSessionReachedDifferentSessions()
+        {
+            long sessionCacheMaxSize = 2;
+            CryptoPolicy policy = BasicExpiringCryptoPolicy.NewBuilder()
+                .WithKeyExpirationDays(1)
+                .WithRevokeCheckMinutes(30)
+                .WithCanCacheSessions(true)
+                .WithSessionCacheMaxSize(sessionCacheMaxSize)
+                .Build();
+
+            using (SessionFactory factory = SessionFactory.NewBuilder(TestProductId, TestServiceId)
+                .WithInMemoryMetastore()
+                .WithCryptoPolicy(policy)
+                .WithStaticKeyManagementService(TestMasterKey)
+                .Build())
+            {
+                int numThreads = 100;
+                int numRequests = 100;
+
+                // Get the current settings and try to force minWorkers
+                ThreadPool.GetMinThreads(out _, out var currentMinIOC);
+                Assert.True(ThreadPool.SetMinThreads(numThreads, currentMinIOC));
+
+                long completedTasks = 0;
+                Parallel.ForEach(Enumerable.Range(0, numRequests), i =>
+                {
+                    using (Session<byte[], byte[]> sessionBytes = factory.GetSessionBytes(TestPartitionId + i))
+                    {
+                        byte[] payload = { 0, 1, 2, 3, 4, 5, 6, 7 };
+                        byte[] drr = sessionBytes.Encrypt(payload);
+                        byte[] decryptedPayload = sessionBytes.Decrypt(drr);
+
+                        Assert.Equal(payload, decryptedPayload);
+                        Interlocked.Increment(ref completedTasks);
+                    }
+                });
+
+                // Wait for all threads to complete
+                Assert.Equal(numRequests, completedTasks);
```

Comment:
Reckon that, like Java, we cannot reliably assert the cache size here? If so, might be worth adding a comment about the same. If not, verify cache size here.